File: patch_ui.js
const fs = require('fs');
let c = fs.readFileSync('frontend/src/pages/HotelDetail.jsx', 'utf8');

const uiStart = c.indexOf('<div className="flex flex-wrap gap-2">');
const uiEnd = c.indexOf('<div style={{ height: \'500px\'');
const uiCode = c.substring(uiStart, uiEnd);

const newUiCode = `<div className="flex flex-col gap-4 mb-4">
                  <div className="flex items-center gap-2">
                    <span className="text-sm font-semibold text-slate-700 dark:text-slate-300">Trip Type:</span>
                    <div className="flex flex-wrap gap-2">
                      {['family', 'business', 'couple', 'solo'].map(t => (
                        <button 
                          key={t}
                          type="button" 
                          onClick={() => setTripType(t)} 
                          className={\`px-3 py-1 text-xs rounded-full border \${tripType === t ? 'bg-brand-600 text-white border-brand-500' : 'bg-slate-50 text-slate-600 border-slate-300 hover:bg-slate-100 dark:bg-dark-950/60 dark:text-slate-400 dark:border-slate-700'}\`}
                        >
                          {t.charAt(0).toUpperCase() + t.slice(1)}
                        </button>
                      ))}
                    </div>
                  </div>
                  
                  <div className="flex items-center gap-2">
                    <span className="text-sm font-semibold text-slate-700 dark:text-slate-300">Category Filter:</span>
                    <div className="flex flex-wrap gap-2">
                      <button type="button" onClick={() => setMapFilter('all')} className={\`px-3 py-1 text-xs rounded-full border \${mapFilter === 'all' ? 'bg-slate-800 text-white border-slate-700 dark:bg-slate-100 dark:text-slate-900' : 'bg-slate-50 text-slate-600 border-slate-300 hover:bg-slate-100 dark:bg-dark-950/60 dark:text-slate-400 dark:border-slate-700'}\`}>All Categories</button>
                    </div>
                  </div>
                </div>

                {isFetchingPlaces && (
                  <div className="flex items-center gap-2 text-xs text-slate-500 dark:text-slate-400">
                    <Loader2 className="w-4 h-4 animate-spin text-brand-500" />
                    <span>Fetching real-time OpenStreetMap / Overpass API surrounding attractions...</span>
                  </div>
                )}
              </div>
              
              <div className="grid grid-cols-1 lg:grid-cols-3 gap-6">
                <div className="lg:col-span-2">
                  <div style={{ height: '500px', width: '100%', borderRadius: '8px', overflow: 'hidden' }} className="border border-slate-200 dark:border-slate-700">
`;

c = c.replace(uiCode, newUiCode);

// Add the list to the right of the map
const mapEndIdx = c.indexOf('</MapContainer>');
const mapBlockEnd = c.indexOf('</div>', mapEndIdx); // the closing div of the map
const mapBlockEnd2 = c.indexOf('</div>', mapBlockEnd + 6); // the closing div of the panel?
// Actually just replace </MapContainer>\n              </div>
const mapEndToReplace = '</MapContainer>\n              </div>';
const newMapEnd = `</MapContainer>
                  </div>
                </div>
                
                <div className="lg:col-span-1 flex flex-col gap-3 max-h-[500px] overflow-y-auto pr-2 custom-scrollbar">
                  <h4 className="font-bold text-slate-900 dark:text-white sticky top-0 bg-white dark:bg-dark-900 py-2 z-10 border-b border-slate-100 dark:border-slate-800">Nearby Places List</h4>
                  {nearbyPlaces.length === 0 && !isFetchingPlaces && (
                    <p className="text-sm text-slate-500">No places found nearby.</p>
                  )}
                  {nearbyPlaces.filter(p => mapFilter === 'all' || p.category === mapFilter).map(place => (
                    <div key={place.id} className="p-3 border border-slate-200 dark:border-slate-800 rounded-lg flex flex-col gap-1 hover:border-brand-300 transition-colors bg-slate-50 dark:bg-dark-950/50">
                      <div className="font-bold text-sm text-slate-800 dark:text-slate-200">{place.name}</div>
                      <div className="flex justify-between items-center text-xs">
                        <span className="px-2 py-0.5 rounded text-white font-semibold text-[10px]" style={{backgroundColor: place.color}}>{place.displayCategory || place.category}</span>
                        <span className="text-slate-500 font-medium">{place.distanceKm?.toFixed(2)} km</span>
                      </div>
                    </div>
                  ))}
                </div>
              </div>`;

c = c.replace(mapEndToReplace, newMapEnd);

// Fix the map marker rendering
// Old marker rendering uses distance calculation which is now provided by backend
c = c.replace(/const dist = getDistanceFromLatLonInKm\(hotel\.latitude, hotel\.longitude, place\.lat, place\.lon\);/g, "const dist = place.distanceKm || 0;");

// Update filter condition for markers
c = c.replace(/\{nearbyPlaces\.filter\(p => mapFilter === 'all' \|\| p\.group === mapFilter\)\.map\(\(place\) => \{/g, "{nearbyPlaces.filter(p => mapFilter === 'all' || p.category === mapFilter).map((place) => {");

// Update lat/lon to match backend which sends latitude/longitude instead of lat/lon
c = c.replace(/center=\{\[place\.lat, place\.lon\]\}/g, "center={[place.latitude, place.longitude]}");
c = c.replace(/getDistanceFromLatLonInKm\(hotel\.latitude, hotel\.longitude, place\.latitude, place\.longitude\)/g, "place.distanceKm || 0");

fs.writeFileSync('frontend/src/pages/HotelDetail.jsx', c);

File: patch_ui.py
import re

with open("frontend/src/components/BenchmarkingView.jsx", "r", encoding="utf-8") as f:
    content = f.read()

old_return = """  return (
    <div className="space-y-6">
      <div className="flex flex-col sm:flex-row justify-between items-start sm:items-center gap-4">
        <div>
          <h3 className="text-xl font-bold text-slate-900">Competitor Benchmarking</h3>"""

new_return = """  return (
    <div className="space-y-6">
      
      {/* Date Range Filter UI */}
      <div className="bg-white dark:bg-dark-900 p-6 rounded-2xl border border-slate-200 dark:border-slate-800 shadow-sm">
        <h4 className="text-sm font-bold text-slate-800 dark:text-slate-100 mb-4">Comparison Period</h4>
        <div className="flex flex-col md:flex-row gap-4 items-end">
          <div className="flex-1">
            <label className="block text-xs font-semibold text-slate-600 dark:text-slate-300 mb-1">Start Date</label>
            <input 
              type="date" 
              value={startDate} 
              onChange={e => { setStartDate(e.target.value); setActivePreset(null); }} 
              className="w-full px-4 py-2 bg-slate-50 dark:bg-dark-800 border border-slate-200 dark:border-slate-700 rounded-xl focus:outline-none focus:ring-2 focus:ring-brand-500 text-slate-900 dark:text-white" 
            />
          </div>
          <div className="flex-1">
            <label className="block text-xs font-semibold text-slate-600 dark:text-slate-300 mb-1">End Date</label>
            <input 
              type="date" 
              value={endDate} 
              onChange={e => { setEndDate(e.target.value); setActivePreset(null); }} 
              className="w-full px-4 py-2 bg-slate-50 dark:bg-dark-800 border border-slate-200 dark:border-slate-700 rounded-xl focus:outline-none focus:ring-2 focus:ring-brand-500 text-slate-900 dark:text-white" 
            />
          </div>
          <div className="flex gap-2">
            <button 
              onClick={() => handleApplyFilter()} 
              disabled={loading} 
              className="px-5 py-2 bg-brand-600 text-white rounded-xl text-sm font-bold hover:bg-brand-700 transition disabled:opacity-50"
            >
              {loading ? 'Applying...' : 'Apply'}
            </button>
            <button 
              onClick={() => applyPreset('summer')} 
              disabled={loading} 
              className="px-4 py-2 bg-amber-100 dark:bg-amber-900/30 text-amber-800 dark:text-amber-200 rounded-xl text-sm font-bold hover:bg-amber-200 dark:hover:bg-amber-900/50 transition"
            >
              Summer Season
            </button>
            <button 
              onClick={() => applyPreset('winter')} 
              disabled={loading} 
              className="px-4 py-2 bg-blue-100 dark:bg-blue-900/30 text-blue-800 dark:text-blue-200 rounded-xl text-sm font-bold hover:bg-blue-200 dark:hover:bg-blue-900/50 transition"
            >
              Winter Season
            </button>
          </div>
        </div>
        {period && (
          <div className="mt-4 pt-4 border-t border-slate-100 dark:border-slate-800">
            <p className="text-sm text-brand-600 dark:text-brand-400 font-semibold">
              Comparison Period: {period.start_date} &rarr; {period.end_date}
              {activePreset && ` | Season: ${activePreset}`}
            </p>
          </div>
        )}
      </div>

      <div className="flex flex-col sm:flex-row justify-between items-start sm:items-center gap-4">
        <div>
          <h3 className="text-xl font-bold text-slate-900 dark:text-white">Competitor Benchmarking</h3>"""

if old_return in content:
    content = content.replace(old_return, new_return)
    print("Main return patched.")
else:
    print("Main return NOT found.")

old_early_return = """  if (marketAverage.total_competitors === 0) {
    return (
      <div className="space-y-6">
        <h3 className="text-xl font-bold text-slate-900">Competitor Benchmarking</h3>"""

new_early_return = """  if (marketAverage.total_competitors === 0) {
    return (
      <div className="space-y-6">
        {/* Date Range Filter UI */}
        <div className="bg-white dark:bg-dark-900 p-6 rounded-2xl border border-slate-200 dark:border-slate-800 shadow-sm">
          <h4 className="text-sm font-bold text-slate-800 dark:text-slate-100 mb-4">Comparison Period</h4>
          <div className="flex flex-col md:flex-row gap-4 items-end">
            <div className="flex-1">
              <label className="block text-xs font-semibold text-slate-600 dark:text-slate-300 mb-1">Start Date</label>
              <input 
                type="date" 
                value={startDate} 
                onChange={e => { setStartDate(e.target.value); setActivePreset(null); }} 
                className="w-full px-4 py-2 bg-slate-50 dark:bg-dark-800 border border-slate-200 dark:border-slate-700 rounded-xl focus:outline-none focus:ring-2 focus:ring-brand-500 text-slate-900 dark:text-white" 
              />
            </div>
            <div className="flex-1">
              <label className="block text-xs font-semibold text-slate-600 dark:text-slate-300 mb-1">End Date</label>
              <input 
                type="date" 
                value={endDate} 
                onChange={e => { setEndDate(e.target.value); setActivePreset(null); }} 
                className="w-full px-4 py-2 bg-slate-50 dark:bg-dark-800 border border-slate-200 dark:border-slate-700 rounded-xl focus:outline-none focus:ring-2 focus:ring-brand-500 text-slate-900 dark:text-white" 
              />
            </div>
            <div className="flex gap-2">
              <button 
                onClick={() => handleApplyFilter()} 
                disabled={loading} 
                className="px-5 py-2 bg-brand-600 text-white rounded-xl text-sm font-bold hover:bg-brand-700 transition disabled:opacity-50"
              >
                {loading ? 'Applying...' : 'Apply'}
              </button>
              <button 
                onClick={() => applyPreset('summer')} 
                disabled={loading} 
                className="px-4 py-2 bg-amber-100 dark:bg-amber-900/30 text-amber-800 dark:text-amber-200 rounded-xl text-sm font-bold hover:bg-amber-200 dark:hover:bg-amber-900/50 transition"
              >
                Summer Season
              </button>
              <button 
                onClick={() => applyPreset('winter')} 
                disabled={loading} 
                className="px-4 py-2 bg-blue-100 dark:bg-blue-900/30 text-blue-800 dark:text-blue-200 rounded-xl text-sm font-bold hover:bg-blue-200 dark:hover:bg-blue-900/50 transition"
              >
                Winter Season
              </button>
            </div>
          </div>
          {period && (
            <div className="mt-4 pt-4 border-t border-slate-100 dark:border-slate-800">
              <p className="text-sm text-brand-600 dark:text-brand-400 font-semibold">
                Comparison Period: {period.start_date} &rarr; {period.end_date}
                {activePreset && ` | Season: ${activePreset}`}
              </p>
            </div>
          )}
        </div>

        <h3 className="text-xl font-bold text-slate-900">Competitor Benchmarking</h3>"""

if old_early_return in content:
    content = content.replace(old_early_return, new_early_return)
    print("Early return patched.")
else:
    print("Early return NOT found.")

with open("frontend/src/components/BenchmarkingView.jsx", "w", encoding="utf-8") as f:
    f.write(content)

